File: rest-api/src/main/java/com/kt/kol/api/comment/service/CommentService.java
package com.kt.kol.api.comment.service;

import com.kt.kol.api.comment.model.Comment;
import com.kt.kol.api.comment.model.CommentDTO;
import com.kt.kol.api.comment.repository.CommentRepository;
import com.kt.kol.api.post.repository.PostRepository;
import com.kt.kol.api.user.repository.UserRepository;
import com.kt.kol.common.exception.BusinessException;
import com.kt.kol.common.model.PageDTO;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import reactor.core.publisher.Mono;

import java.time.LocalDateTime;
import java.util.List;

@Slf4j
@Service
@RequiredArgsConstructor
public class CommentService {

    private final CommentRepository commentRepository;
    private final PostRepository postRepository;
    private final UserRepository userRepository;

    public Mono<PageDTO<CommentDTO>> findCommentsByPostIdPaged(Long postId, int page, int size) {
        long offset = (long) page * size;

        Mono<Long> countMono = commentRepository.countByPostId(postId);
        Mono<List<CommentDTO>> dataMono = commentRepository.findByPostIdPage(postId, offset, size)
                .flatMap(this::enrichCommentWithUser)
                .collectList();

        return Mono.zip(countMono, dataMono)
                .map(tuple -> {
                    Long totalElements = tuple.getT1();
                    List<CommentDTO> content = tuple.getT2();

                    if (totalElements == 0) {
                        return PageDTO.empty(page, size);
                    }
                    return PageDTO.of(content, page, size, totalElements);
                });
    }

    public Mono<PageDTO<CommentDTO>> findRootCommentsByPostIdPaged(Long postId, int page, int size) {
        long offset = (long) page * size;

        Mono<Long> countMono = commentRepository.countRootCommentsByPostId(postId);
        Mono<List<CommentDTO>> dataMono = commentRepository.findRootCommentsByPostIdPage(postId, offset, size)
                .flatMap(this::enrichCommentWithUser)
                .collectList();

        return Mono.zip(countMono, dataMono)
                .map(tuple -> {
                    Long totalElements = tuple.getT1();
                    List<CommentDTO> content = tuple.getT2();

                    if (totalElements == 0) {
                        return PageDTO.empty(page, size);
                    }
                    return PageDTO.of(content, page, size, totalElements);
                });
    }

    public Mono<PageDTO<CommentDTO>> findRepliesByCommentIdPaged(Long commentId, int page, int size) {
        long offset = (long) page * size;

        Mono<Long> countMono = commentRepository.countByParentId(commentId);
        Mono<List<CommentDTO>> dataMono = commentRepository.findByParentIdPage(commentId, offset, size)
                .flatMap(this::enrichCommentWithUser)
                .collectList();

        return Mono.zip(countMono, dataMono)
                .map(tuple -> {
                    Long totalElements = tuple.getT1();
                    List<CommentDTO> content = tuple.getT2();

                    if (totalElements == 0) {
                        return PageDTO.empty(page, size);
                    }
                    return PageDTO.of(content, page, size, totalElements);
                });
    }

    @Transactional
    public Mono<CommentDTO> createComment(CommentDTO commentDTO) {
        return validateCommentCreation(commentDTO)
                .then(Mono.defer(() -> {
                    Comment comment = toEntity(commentDTO);
                    comment.setCreatedAt(LocalDateTime.now());
                    comment.setUpdatedAt(LocalDateTime.now());
                    return commentRepository.save(comment);
                }))
                .flatMap(this::enrichCommentWithUser)
                .doOnSuccess(comment -> log.info("Comment created for post: {}", comment.postId()));
    }

    @Transactional
    public Mono<CommentDTO> updateComment(Long id, CommentDTO commentDTO) {
        return commentRepository.findById(id)
                .switchIfEmpty(Mono.error(new BusinessException("COMMENT_NOT_FOUND", "댓글을 찾을 수 없습니다.")))
                .flatMap(existingComment -> {
                    existingComment.setContent(commentDTO.content());
                    existingComment.setUpdatedAt(LocalDateTime.now());
                    return commentRepository.save(existingComment);
                })
                .flatMap(this::enrichCommentWithUser)
                .doOnSuccess(comment -> log.info("Comment updated: {}", comment.id()));
    }

    @Transactional
    public Mono<Void> deleteComment(Long id) {
        return commentRepository.findById(id)
                .switchIfEmpty(Mono.error(new BusinessException("COMMENT_NOT_FOUND", "댓글을 찾을 수 없습니다.")))
                .flatMap(comment -> {
                    return commentRepository.countByParentId(id)
                            .flatMap(replyCount -> {
                                if (replyCount > 0) {
                                    return Mono.error(new BusinessException("HAS_REPLIES", "답글이 있는 댓글은 삭제할 수 없습니다."));
                                }
                                return commentRepository.delete(comment);
                            });
                })
                .doOnSuccess(v -> log.info("Comment deleted: {}", id));
    }

    public Mono<Long> countCommentsByPostId(Long postId) {
        return commentRepository.countByPostId(postId);
    }

    private Mono<Void> validateCommentCreation(CommentDTO commentDTO) {
        return postRepository.existsById(commentDTO.postId())
                .flatMap(exists -> {
                    if (!exists) {
                        return Mono.error(new BusinessException("POST_NOT_FOUND", "게시글을 찾을 수 없습니다."));
                    }
                    return userRepository.existsById(commentDTO.userId());
                })
                .flatMap(exists -> {
                    if (!exists) {
                        return Mono.error(new BusinessException("USER_NOT_FOUND", "사용자를 찾을 수 없습니다."));
                    }
                    if (commentDTO.parentId() != null) {
                        return commentRepository.existsById(commentDTO.parentId())
                                .flatMap(parentExists -> {
                                    if (!parentExists) {
                                        return Mono.error(
                                                new BusinessException("PARENT_COMMENT_NOT_FOUND", "부모 댓글을 찾을 수 없습니다."));
                                    }
                                    return Mono.empty();
                                });
                    }
                    return Mono.empty();
                });
    }

    private Mono<CommentDTO> enrichCommentWithUser(Comment comment) {
        return userRepository.findById(comment.getUserId())
                .map(user -> toDTO(comment, user.getUsername()))
                .defaultIfEmpty(toDTO(comment, "Unknown"));
    }

    private CommentDTO toDTO(Comment comment, String username) {
        return new CommentDTO(
                comment.getId(),
                comment.getContent(),
                comment.getPostId(),
                comment.getUserId(),
                username,
                comment.getParentId(),
                comment.getCreatedAt(),
                comment.getUpdatedAt());
    }

    private Comment toEntity(CommentDTO dto) {
        return Comment.builder()
                .content(dto.content())
                .postId(dto.postId())
                .userId(dto.userId())
                .parentId(dto.parentId())
                .build();
    }
}

File: rest-api/src/main/kotlin/com/kt/kol/api/comment/service/CommentService.kt
package com.kt.kol.api.comment.service

import com.kt.kol.api.comment.model.Comment
import com.kt.kol.api.comment.model.CommentDTO
import com.kt.kol.api.comment.repository.CommentRepository
import com.kt.kol.api.post.repository.PostRepository
import com.kt.kol.api.user.repository.UserRepository
import com.kt.kol.common.exception.BusinessException
import com.kt.kol.common.model.PageDTO
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.toList
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime

private val log = LoggerFactory.getLogger(CommentService::class.java)

@Service
class CommentService(
    private val commentRepository: CommentRepository,
    private val postRepository: PostRepository,
    private val userRepository: UserRepository
) {

    suspend fun findCommentsByPostIdPaged(postId: Long, page: Int, size: Int): PageDTO<CommentDTO> {
        val offset = page.toLong() * size

        val totalElements = commentRepository.countByPostId(postId)
        val content = commentRepository.findByPostIdPage(postId, offset, size)
            .map { enrichCommentWithUser(it) }
            .toList()

        return if (totalElements == 0L) {
            PageDTO.empty(page, size)
        } else {
            PageDTO.of(content, page, size, totalElements)
        }
    }

    suspend fun findRootCommentsByPostIdPaged(postId: Long, page: Int, size: Int): PageDTO<CommentDTO> {
        val offset = page.toLong() * size

        val totalElements = commentRepository.countRootCommentsByPostId(postId)
        val content = commentRepository.findRootCommentsByPostIdPage(postId, offset, size)
            .map { enrichCommentWithUser(it) }
            .toList()

        return if (totalElements == 0L) {
            PageDTO.empty(page, size)
        } else {
            PageDTO.of(content, page, size, totalElements)
        }
    }

    suspend fun findRepliesByCommentIdPaged(commentId: Long, page: Int, size: Int): PageDTO<CommentDTO> {
        val offset = page.toLong() * size

        val totalElements = commentRepository.countByParentId(commentId)
        val content = commentRepository.findByParentIdPage(commentId, offset, size)
            .map { enrichCommentWithUser(it) }
            .toList()

        return if (totalElements == 0L) {
            PageDTO.empty(page, size)
        } else {
            PageDTO.of(content, page, size, totalElements)
        }
    }

    @Transactional
    suspend fun createComment(commentDTO: CommentDTO): CommentDTO {
        validateCommentCreation(commentDTO)

        val comment = toEntity(commentDTO).apply {
            createdAt = LocalDateTime.now()
            updatedAt = LocalDateTime.now()
        }

        val savedComment = commentRepository.save(comment)
        val result = enrichCommentWithUser(savedComment)
        log.info("Comment created for post: ${result.postId}")
        return result
    }

    @Transactional
    suspend fun updateComment(id: Long, commentDTO: CommentDTO): CommentDTO {
        val existingComment = commentRepository.findById(id)
            ?: throw BusinessException("COMMENT_NOT_FOUND", "댓글을 찾을 수 없습니다.")

        existingComment.apply {
            content = commentDTO.content
            updatedAt = LocalDateTime.now()
        }

        val savedComment = commentRepository.save(existingComment)
        val result = enrichCommentWithUser(savedComment)
        log.info("Comment updated: ${result.id}")
        return result
    }

    @Transactional
    suspend fun deleteComment(id: Long) {
        val comment = commentRepository.findById(id)
            ?: throw BusinessException("COMMENT_NOT_FOUND", "댓글을 찾을 수 없습니다.")

        val replyCount = commentRepository.countByParentId(id)
        if (replyCount > 0) {
            throw BusinessException("HAS_REPLIES", "답글이 있는 댓글은 삭제할 수 없습니다.")
        }

        commentRepository.delete(comment)
        log.info("Comment deleted: $id")
    }

    suspend fun countCommentsByPostId(postId: Long): Long {
        return commentRepository.countByPostId(postId)
    }

    private suspend fun validateCommentCreation(commentDTO: CommentDTO) {
        val postExists = postRepository.existsById(commentDTO.postId!!)
        if (!postExists) {
            throw BusinessException("POST_NOT_FOUND", "게시글을 찾을 수 없습니다.")
        }

        val userExists = userRepository.existsById(commentDTO.userId!!)
        if (!userExists) {
            throw BusinessException("USER_NOT_FOUND", "사용자를 찾을 수 없습니다.")
        }

        commentDTO.parentId?.let { parentId ->
            val parentExists = commentRepository.existsById(parentId)
            if (!parentExists) {
                throw BusinessException("PARENT_COMMENT_NOT_FOUND", "부모 댓글을 찾을 수 없습니다.")
            }
        }
    }

    private suspend fun enrichCommentWithUser(comment: Comment): CommentDTO {
        val username = try {
            userRepository.findById(comment.userId!!)?.username ?: "Unknown"
        } catch (e: Exception) {
            "Unknown"
        }
        return toDTO(comment, username)
    }

    private fun toDTO(comment: Comment, username: String): CommentDTO {
        return CommentDTO(
            id = comment.id,
            content = comment.content,
            postId = comment.postId,
            userId = comment.userId,
            username = username,
            parentId = comment.parentId,
            createdAt = comment.createdAt,
            updatedAt = comment.updatedAt
        )
    }

    private fun toEntity(dto: CommentDTO): Comment {
        return Comment(
            content = dto.content,
            postId = dto.postId,
            userId = dto.userId,
            parentId = dto.parentId
        )
    }
}